ValveAnnulusAnalysis/HeartValveWidgets/ValveSequenceBrowserWidget.py

```diff
@@ -0,0 +1,267 @@
+        self.valveVolumeBrowserNode.AddObserver(vtk.vtkCommand.ModifiedEvent, self.updateGUIFromValveVolumeBrowser)
+
+  def _removeValveVolumeBrowserObserver(self):
+    if self.valveVolumeBrowserNode and self.valveVolumeBrowserNodeObserver:
+      self.valveVolumeBrowserNode.RemoveObserver(self.valveVolumeBrowserNodeObserver)
+      self.valveVolumeBrowserNodeObserver = None
+
+  def onCardiacCyclePhaseChanged(self):
+    if self.valveModel is None:
+      return
+    self.valveModel.setCardiacCyclePhase(self.ui.cardiacCyclePhaseSelector.currentText)
+
+  def onAddTimePointButtonClicked(self):
+    volumeSequenceIndex, volumeSequenceIndexValue = self.valveBrowser.getDisplayedValveVolumeSequenceIndexAndValue()
+    heartValveSequenceIndex = self.valveBrowser.heartValveSequenceNode.GetItemNumberFromIndexValue(volumeSequenceIndexValue)
+
+    if heartValveSequenceIndex >= 0: # item for timepoint exists
+      self.valveBrowser.valveBrowserNode.SetSelectedItemNumber(heartValveSequenceIndex)
+    else:
+      logging.info(f"Add time point")
+      if not volumeSequenceIndexValue:
+        raise RuntimeError("Failed to add time point, could not get volume sequence")
+      self.valveBrowser.addTimePoint(volumeSequenceIndexValue)
+      heartValveNode = self.valveBrowser.heartValveNode if self.valveBrowser else None
+      self.setHeartValveNode(heartValveNode)
+
+    self.updateGUIFromValveVolumeBrowser()
+
+  def onRemoveTimePointButtonClicked(self):
+    itemIndex, indexValue = self.valveBrowser.getDisplayedHeartValveSequenceIndexAndValue()
+    if indexValue is not None:
+      # TODO: add confirm dialog
+      self.valveBrowser.removeTimePoint(indexValue)
+
+    self.updateGUIFromHeartValveNode()
+
+  def disableGUI(self):
+    self.ui.addTimePointButton.text = "Add volume"
+    self._setCardiacCycleIndex(0)
+    self.ui.setEnabled(False)
+
+  def updateGUIFromValveVolumeBrowser(self, unusedArg1=None, unusedArg2=None, unusedArg3=None):
+    if not self.valveVolumeBrowserNode or self.valveVolumeBrowserNode.GetPlaybackActive():
+      self.disableGUI()
+      return
+
+    seqIndex, seqIndexValue = self.valveBrowser.getDisplayedValveVolumeSequenceIndexAndValue()
+    hasTimepoint = seqIndexValue and self.valveBrowser.heartValveSequenceNode.GetItemNumberFromIndexValue(seqIndexValue) >= 0
+
+    self.ui.setEnabled(self.valveBrowser.valveBrowserNode is not None if self.valveBrowser else False)
+    self.ui.removeTimePointButton.enabled = hasTimepoint
+    self.ui.cardiacCyclePhaseSelector.enabled = hasTimepoint
+
+    t = Template("$action volume (index = $index)")
+    if hasTimepoint:
+      self.ui.addTimePointButton.text = t.substitute(action="Go to", index=seqIndex + 1)
+      heartValveNode = self.valveBrowser.heartValveSequenceNode.GetDataNodeAtValue(seqIndexValue)
+      self._setCardiacCycleIndex(
+        self.ui.cardiacCyclePhaseSelector.findText(heartValveNode.GetAttribute("CardiacCyclePhase"))
```

Comment:
Can we use `self.valveModel.getCardiacCyclePhase()` as in line 236? It would even nicer to have only one place where update of self.ui.cardiacCyclePhaseSelector from MRML is implemented.

Reply:
Agreed.